common/src/main/java/io/netty/util/concurrent/Promise.java

```diff
@@ -59,9 +59,79 @@
      */
     boolean setUncancellable();
 
-    @Override
-    Promise<V> addListener(FutureListener<? super V> listener);
+    /**
+     * Cancel this promise, unless it has already been completed.
+     * <p>
+     * A cancelled promise is considered to be {@linkplain Future#isFailed() failed}.
+     * <p>
+     * If the cancellation was successful it will fail associated futures with a {@link CancellationException}.
+     * <p>
+     * Cancellation will not cause any threads working on the promise to be {@linkplain Thread#interrupt()
+     * interrupted}.
+     *
+     * @return {@code true} if the promise was cancelled by this call, otherwise {@code false}.
+     */
+    boolean cancel();
 
-    @Override
-    <C> Promise<V> addListener(C context, FutureContextListener<? super C, ? super V> listener);
+    /**
+     * Returns {@code true} if and only if the operation was completed successfully.
+     */
+    boolean isSuccess();
+
+    /**
+     * Returns {@code true} if and only if the operation was completed and failed.
+     */
+    boolean isFailed();
+
+    /**
+     * Return {@code true} if this promise has been {@linkplain #cancel() cancelled}.
+     *
+     * @return {@code true} if this promise has been cancelled, otherwise {@code false}.
+     */
+    boolean isCancelled();
+
+    /**
+     * Return {@code true} if this promise has been completed either {@linkplain #setSuccess(Object)
+     * successfully}, {@linkplain #setFailure(Throwable) unsuccessfully}, or through
+     * {@linkplain #cancel() cancellation}.
+     *
+     * @return {@code true} if this promise has completed, otherwise {@code false}.
+     */
+    boolean isDone();
+
+    /**
+     * returns {@code true} if and only if the operation can be cancelled via {@link #cancel()}.
+     */
+    boolean isCancellable();
+
+    /**
+     * Return the successful result of this promise, if any.
+     * If the promise has not yet been completed, then this will throw {@link IllegalStateException}.
+     * If the promise has been cancelled or failed with an exception, then this returns {@code null}.
+     * Note that a promise can also be completed successfully with a {@code null} result.
+     *
+     * @return the result of this promise, if completed successfully.
+     * @throws IllegalStateException if this {@code Future} has not completed yet.
+     */
+    V getNow();
+
+    /**
+     * Returns the cause of the failed operation if the operation has failed.
+     *
+     * @return The cause of the failure, if any. Otherwise {@code null} if succeeded.
+     * @throws IllegalStateException if this {@code Promise} has not completed yet.
+     */
+    Throwable cause();
+
+    /**
+     * Produce a {@link Future} that will be completed upon completion of this promise.
+     *
+     * @return A future instance associated with this promise.
+     */
+    Future<V> toFuture();
```

Comment:
Imho we should either rename this ´asFuture()´ and make it a contract that the same instance must be returned all time or we should rename `asCompletionStage()` to `toCompletionStage()`. Just to keep things consistent

Reply:
Yeah, I think that makes sense, with `asFuture`. Makes it clear that this is a cheap method to call as well.